pkg/checks/so_name.go

```diff
@@ -0,0 +1,279 @@
+		c, err := melange.ReadMelangeConfig(filename)
+		if err != nil {
+			log.Printf("failed to read melange config %s", filename)
+			continue
+		}
+
+		for i := 0; i < len(c.Subpackages); i++ {
+			packagesAndSubpackages[c.Subpackages[i].Name] = apkPackage
+		}
+	}
+	return packagesAndSubpackages
+}
+
+// diff will compare the so name versions between the latest existing apk in a APKINDEX with a newly built local apk
+//
+//nolint:gocritic // hugeParam for entry
+func (o SoNameOptions) diff(newPackageName string, newAPK NewApkPackage, existingPackages map[string]*repository.Package) error {
+	dirExistingApk := os.TempDir()
+	dirNewApk := os.TempDir()
+
+	// read new apk
+	filename := filepath.Join(o.PackagesDir, newAPK.Arch, fmt.Sprintf("%s-%s-r%s.apk", newPackageName, newAPK.Version, newAPK.Epoch))
+	newFile, err := os.Open(filename)
+	if err != nil {
+		return errors.Wrapf(err, "failed to read %s", filename)
+	}
+
+	err = tar.Untar(newFile, dirNewApk)
+	if err != nil {
+		return errors.Wrapf(err, "failed to untar new apk")
+	}
+
+	newSonameFiles, err := o.getSonameFiles(dirNewApk)
+	if err != nil {
+		return errors.Wrapf(err, "error when looking for soname files in new apk")
+	}
+	// if no .so name files, skip
+	if len(newSonameFiles) == 0 {
+		return nil
+	}
+
+	// fetch current latest apk
+	p := existingPackages[newPackageName]
+
+	if p == nil {
+		o.Logger.Printf("no existing package found for %s, skipping so name check", newPackageName)
+		return nil
+	}
+	existingFilename := fmt.Sprintf("%s-%s.apk", p.Name, p.Version)
+	err = o.downloadCurrentAPK(existingFilename, dirExistingApk)
+	if err != nil {
+		return errors.Wrapf(err, "failed to download %s using base URL %s", newPackageName, o.ApkIndexURL)
+	}
+
+	// get any existing so names
+	existingSonameFiles, err := o.getSonameFiles(dirExistingApk)
+	if err != nil {
+		return errors.Wrapf(err, "error when looking for soname files in existing apk")
+	}
+
+	err = o.checkSonamesMatch(existingSonameFiles, newSonameFiles)
+	if err != nil {
+		return errors.Wrapf(err, "soname files differ, this can cause an ABI break.  Existing soname files %s, New soname files %s", strings.Join(existingSonameFiles, ","), strings.Join(newSonameFiles, ","))
+	}
+
+	return nil
+}
+
+//nolint:gocritic // hugeParam for entry
+func (o SoNameOptions) downloadCurrentAPK(newPackageName string, dirCurrentApk string) error {
+	apkURL := strings.ReplaceAll(o.ApkIndexURL, "APKINDEX", newPackageName)
+	resp, err := o.Client.Get(apkURL)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get %s", apkURL)
+	}
+	defer resp.Body.Close()
+
+	// Create the file
+	out, err := os.Create(filepath.Join(dirCurrentApk, newPackageName))
+	if err != nil {
+		return err
+	}
+	defer out.Close()
+
+	// Writer the body to file
+	_, err = io.Copy(out, resp.Body)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+//nolint:gocritic // hugeParam for entry
+func (o SoNameOptions) getSonameFiles(dir string) ([]string, error) {
```

Comment:
Why not use a pointer to avoid the hugeParam issue?

Reply:
only becuase by IDE complained with a warning when making it a pointer
```
Struct SoNameOptions has methods on both value and pointer receivers. Such usage is not recommended by the Go Documentation.
```
I've not looked closely at why, so probably my code.  Will check in the morning.